templated-this.md

```diff
@@ -1,151 +1,315 @@
+screen.
+
+Teaching also becomes easier, as the meaning of what a 'const method' is becomes
+more obvious to students.
+
+One can always obtain the address of the object by taking the address of the
+`this`.
+
+
+### unification with `inline friend` functions
+
+This proposal also makes `this` far less special. In fact, it completely unifies
+`inline friend` functions and class methods, with the differences being:
+
+- the calling convention
+- methods can be virtual
+
+Basically, if the first parameter is called `this`, one can parse and
+instantiate the declaration with exactly the same rules as an `inline friend`,
+except with a calling convention for methods.
+
+
+### `virtual`
+
+We are not entirely sure how `virtual` and value calling conventions would work,
+so perhaps we can disallow that use-case, if the pass-by-value signatures are
+kept in.
+
+Implications for lambdas
+------------------------
+
+Generic lambdas, should they take an `auto&& this` parameter, work according to
+existing rewriting rules: the `auto&& this` is turned into a "universal
+reference" and deduced as if it were inside a
+`template <typename T> auto operator()(T&& this) -> <return type> {...}`.)
+
+### Interplays with capturing `[this]` and `[*this]`
+
+As an explicit parameter, it shadows members of the closure. That said, using
+`this` inside the lambda is useless for anything other than `decltype(this)`,
+since the closure members have no defined names.
+
+TBD: does init-capture obviate all need for `*this`?
+
+### Do we allow `this` in lambdas that decay to a function pointer? 
+
+If the lambda would otherwise decay to a function pointer, `&this` shall have
+the value of that function pointer.
+
+### Does this allow recursion in lambdas?
+
+Yup. You're allowed to call `this(...)`.
+
+
+### Expressions allowed for `this` in lambdas:
+```
+this(...);      // call with appropriate signature
+decltype(this); // evaluates to the type of the lambda with the appropriate
+                // cv-ref qualifiers
+&this;          // the address of either the closure object or function pointer
+std::move(this) // you're allowed to move yourself into an algorithm...
+/* ... and all other things you're allowed to do with the lambda itself. */
+```
+
+TBD: I don't think we should say that you can now refer to the closure object.
+What you can do is deduce the value category of the closure object, and access
+its members in a way appropriate to it.
+
+TBD: Do we believe this?
+
+
+Impact on the Standard
+----------------------
+   
+TBD: A bunch of stuff in section 8.1.5 [expr.prim.lambda].
+TBD: A bunch of stuff in that `this` can appear as the first method parameter.
+
+Acknowledgments
+---------------
+
+Thanks to the following for their help and guidance:
+
+- Louis Dionne
+- Marshal Clowe
+
+References
+----------
+
+- `*this` capture paper
+- ...
+
+We should probably look up the paper that added `*this` capture and why.
+And any other salient papers.
```

Comment:
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0018r3.html (in C++17)
and subsequently:
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0637r0.html

Reply:
Thanks! done.